fix: right click menu issues

microbit-matt-hillsdon opened the pull request:
PR #326 introduced an issue with stale current node when showing the context menu by mouse. It's only possible to fix this by reinstating the updating of curNode from the selection in `getCurNode`.

We also attempt to avoid showing passive focus when using the context menu and widgets (via mouse and keyboard) for consistency with the normal Blockly experience.

You can now no longer get this scenario:

![image](https://github.com/user-attachments/assets/77cc02ae-98d8-41da-be41-95b91287660f)

Fixes an issue clicking on shadow blocks which could previously show an outline around the shadow and parent block. You can now no longer get these scenarios:

![image](https://github.com/user-attachments/assets/c010caf0-8495-491c-9ec6-6d530334d8b3)

![image](https://github.com/user-attachments/assets/12558cf4-978b-43ee-ae15-3f475aabd5b3)

Demo: https://fix-right-click-2.blockly-keyboard-experimentation.pages.dev/

microbit-grace committed:
fix: right click menu issues

PR #326 introduced an issue with stale current node when showing the context
menu by mouse. It's only possible to fix this by reinstating the updating of
curNode from the selection.

We also attempt to avoid showing passive focus when using the context menu and
widgets (via mouse and keyboard) for consistency with the normal Blockly
experience.

Fixes an issue clicking on shadow blocks which could previously show an outline
around the shadow and parent block.

microbit-grace committed:
Merge branch 'main' of https://github.com/google/blockly-keyboard-experimentation into fix-right-click-2

microbit-matt-hillsdon commented on src/navigation.ts:
```diff
@@ -432,6 +440,34 @@ export class Navigation {
+      // style.
+      // https://github.com/google/blockly-samples/issues/2498
+      // https://github.com/google/blockly/issues/8819
+      return;
```
Update now Blockly is upgraded fixing one of the bugs.
We could also just drop this but it's very weird to experience in the demo with the colour picker.

```suggestion
      // Workaround:
      // Skip document.body/null case until this blur bugs is fixed to avoid
      // flipping to passive focus as the user moves their mouse over the
      // colour picker.
      // https://github.com/google/blockly-samples/issues/2498
      return;
```

microbit-matt-hillsdon committed:
chore: remove reference to fixed bug

BenHenning commented on src/navigation.ts:
```diff
@@ -432,6 +440,32 @@ export class Navigation {
+  ) {
+    if (relatedTarget === null) {
+      // Workaround:
+      // Skip document.body/null case until this blur bugs is fixed to avoid
```
```suggestion
      // Skip document.body/null case until this blur bug is fixed to avoid
```

Nit on extra 's'.

microbit-matt-hillsdon replied:
Thanks, fixed in 586a6f0734847fc76e0871619aba6a615f8ae17d

BenHenning commented on src/line_cursor.ts:
```diff
@@ -475,6 +475,43 @@ export class LineCursor extends Marker {
+        selected = selected.getParent();
+      }
+      if (selected) {
+        this.setCurNode(new ASTNode(ASTNode.types.BLOCK, selected), true);
```
Could you perhaps use `Blockly.ASTNode.createBlockNode` here, instead? That's conventional with other locations in the plugin where an `ASTNode` is being created.

microbit-matt-hillsdon replied:
Makes sense. Done as part of 1233f578fd94a0941a8e028a0417d7e83a3eda3f.

BenHenning commented on src/line_cursor.ts:
```diff
@@ -673,22 +709,16 @@ export class LineCursor extends Marker {
-      this.setCurNode(null as never, true);
-    }
+    // This runs too late so the logic to update the selection is in
+    // `getCurNode`.
```
The context here is missing. It makes sense when reading the PR, but this comment by itself post-merge will be hard to understand without blame. Can more context perhaps be added?

Separately, might it be possible to factor out the bits of `getCurNode` which actually update the state into a common helper called both there and here? That might be the cleanest way to indicate what needs to happen in both locations.

microbit-matt-hillsdon replied:
That really helped. 1233f578fd94a0941a8e028a0417d7e83a3eda3f has separate methods for the node -> selection and selection -> node updates and uses the selection -> node one in the listener and `getCurNode`.

microbit-matt-hillsdon replied:
Ultimately I hope https://github.com/google/blockly/issues/8820 might make this less messy

BenHenning replied:
I find the latest changes much cleaner--thanks!

BenHenning reviewed (approved):
Thanks @microbit-matt-hillsdon! I think the fixes make sense. I'm happy to approve the PR, though I did have a few comments on clarity. Feel free to self-resolve and merge, or assign me back if you'd like me to take another review pass.

microbit-matt-hillsdon committed:
fix: stale curNode when clicking the workspace

Due to an unintentially commented out line, clicking the workspace wouldn't
clear the current node so e.g. delete would act on a block that didn't have a
visual selection.

Re-enabling that required revisiting shadow blocks to prevent getCurNode
unexpectedly clearing based on a null selection. We now always set/use the
selection for shadow blocks and take care not to move the cursor unexpectedly
as it can be in two positions for a shadow block selection.

microbit-matt-hillsdon committed:
fix: null node case for action menu

microbit-matt-hillsdon committed:
chore: typo in comment

microbit-matt-hillsdon commented on src/line_cursor.ts:
```diff
@@ -475,6 +475,43 @@ export class LineCursor extends Marker {
+      curNode?.getType() === Blockly.ASTNode.types.BLOCK
+    ) {
+      // Selection says our curNode is not selected anymore.
+      // this.setCurNode(null as never, true);
```
Oops, this is definitely a problem. Re-enabling it threw up a shadow block issue addressed in 1233f578fd94a0941a8e028a0417d7e83a3eda3f

BenHenning replied:
I actually missed this commented out bit the first time around--great catch and thanks for fixing it in the follow-up!

BenHenning reviewed (approved):
Thanks @microbit-matt-hillsdon. Erm, good point on the merging bit. This looks good to me!
